content/en/blog/2024/integrate-datadog-traces-into-otel-stack/index.md

```diff
@@ -0,0 +1,371 @@
+  OpenTelemetry TraceID;
+- `set(trace_id.string, cache["combined_trace_id"]) where cache["combined_trace_id"] != nil`
+  and finally we set the `trace_id` for this span with the string concatenation.
+
+To see it working, let’s run our example again, with a different
+`docker compose` file and then execute the same API call as before:
+
+```bash
+> docker compose -f ./docker-compose.step2.yaml up -d
+
+> curl http://localhost:8080/remotehello
+# it returns:
+# {"hello":"world!","from":"quick-start-api-otel","to":"quick-start-api-datadog"}
+```
+
+When we enter into the Jaeger UI we see that our problem was partially solved.
+Now, we have a trace propagated between the Datadog-instrumented API and the
+OpenTelemetry-instrumented API, however, all child spans generated by Datadog
+are segregated in a different trace, defined only as the Lower TraceID. This
+happens because `ddtrace` only sends a `_dd.p.tid` attribute for the first span
+generated internally, which makes our transform statements skip the child spans.
+
+![Partially disjointed OTelSDK-instrumented trace on Jaeger](./step2-otel-api.png)
+
+![Partially disjointed Datadog-instrumented trace on Jaeger](./step2-datadog-api.png)
+
+### Patching Datadog Trace to Send the Upper TraceID to Every Child Span
+
+Since the OpenTelemetry Collector was
+[designed to process spans considering distributed systems](https://opentelemetry.io/docs/collector/scaling/),
+we don't have a way to maintain an internal state to replicate the `_dd.p.tid`
+attribute for the child spans received by the `datadog` receiver.
+
+We will solve this problem directly on the Datadog-instrumented API, applying a
+minor patch to `ddtrace` to replicate `_dd.p.tid` to all child spans. In the
+`ddtrace` Ruby version, we were able to do that by changing the trace
+serialization to send this data as shown
+[here](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/quick_start_api_datadog/config/initializers/datadog.rb#L20):
+
+```ruby
+module Datadog
+  module Tracing
+    module Transport
+      class SerializableTrace
+        def to_msgpack(packer = nil)
+          if ENV.has_key?('INJECT_UPPER_TRACE_ID')
+            return trace.spans.map { |s| SerializableSpan.new(s) }.to_msgpack(packer)
+          end
+
+          upper_trace_id = trace.spans.find { |span| span.meta.has_key?('_dd.p.tid') }.meta['_dd.p.tid']
+          trace.spans.each do |span|
+            span.meta["propagation.upper_trace_id"] = upper_trace_id
+          end
+
+          trace.spans.map { |s| SerializableSpan.new(s) }.to_msgpack(packer)
+        end
+      end
+    end
+  end
+end
+```
+
+Since it is a customization on the traces, we opted to grab the `_dd.p.tid`
+attribute and inject it in each span as the `propagation.upper_trace_id`
+attribute. We
+[changed](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/collector/collector.config.step3.yaml#L18)
```

Comment:
```suggestion
Since it is a customization on the traces, you can opt to grab the `_dd.p.tid`
attribute and inject it in each span as the `propagation.upper_trace_id`
attribute. Then you can
[change](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/collector/collector.config.step3.yaml#L18)
```